### PR TITLE
Typescript

### LINKED ISSUE #786: Typescript
It's a serious amount of work, but we should add Typescript to keep our app maintainable.
Hopefully it won't have any performance impact or new bugs, but we should do it anyway.

React Native supports Typescript without any harm since 0.57.

We need to do it in baby steps to guarantee not to break anything.
- [ ] Setup: it shouldn't be hard, but we need to make sure everything is working afterwards (app, jest, storybook, detox, release mode, ci)
- [ ] Convert OnboardingView: this is one of the smallest views we have and it's the first one in the navigation stack. Easier to migrate and to test
- [ ] Convert other files module by module (Auth, Rooms, Room, etc).

### PR DESCRIPTION
**Pull request by G2Jose**
@RocketChat/ReactNative

Closes #786 

- Sets up typescript infrastructure. See https://facebook.github.io/react-native/docs/typescript for more details
  - Add a `tsconfig.json` file
  - Update eslint config to support `.ts` and `.tsx` files
  - Update jest config to support typescript files
- Add type definitions from `@types` for
  - `i18n-js`
  - `react`
  - `react-native`
  - `jest`
  - `react-redux`
- Create a `global.types.ts` file exposing common types that'll get used throughout the codebase
- Migrate over OnboardingView and components within to typescript

js files within the project can now be renamed to `.ts`. If they contain JSX, they can be renamed to `.tsx`

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -1,139 +1,143 @@
 module.exports = {
```
Comment (G2Jose) [inline .eslintrc.js]: Looks like this file got auto-formatted. Happy to revert if easier to review

File: .eslintrc.js [hunk 2]
```diff
@@ -1,139 +1,143 @@
...
-		"sourceType": "module",
-		"ecmaVersion": 2017,
-		"ecmaFeatures": {
-			"experimentalObjectRestSpread" : true,
-			"jsx": true,
-			"legacyDecorators": true
+	settings: {
+		'import/resolver': {
+			node: {
+				extensions: ['.js', '.ios.js', '.android.js', '.ts', '.tsx']
```
Comment (G2Jose) [inline .eslintrc.js]: This line has changed

File: .eslintrc.js [hunk 3]
```diff
@@ -1,139 +1,143 @@
...
-        "import",
-    	"react-native"
-	],
-	"env": {
-		"browser": true,
-		"commonjs": true,
-		"es6": true,
-		"node": true,
-		"jquery": true
+	parser: '@typescript-eslint/parser',
```
Comment (G2Jose) [inline .eslintrc.js]: And this

**Comment by diegolmello** [comment]
Nice job! I’ll take a look on it soon! Thank you 👏👏👏

File: .eslintrc.js (hunk 1 above)
Comment (diegolmello) [reply .eslintrc.js]: @G2Jose Yep. Can you revert it? Thanks

Comment (G2Jose) [reply .eslintrc.js]: Done!

**Comment by diegolmello** [comment]
Closed in favor of #3279

**Pull request closed without merging**